src/mpid/ch4/generic/gpu/mpidig_gpu_util.h

```diff
@@ -0,0 +1,182 @@
+        void *gpu_stage_buf_ = MPIDIG_GPU_REQUEST(req, stage_buf);          \
+        if (gpu_stage_buf_) {                                               \
+            if (req->kind == MPIR_REQUEST_KIND__SEND)                       \
+                MPIDIG_gpu_unstage_send_buf(gpu_stage_buf_);                \
+            else if (req->kind == MPIR_REQUEST_KIND__RECV)                  \
+                MPIDIG_gpu_unstage_recv_buf(gpu_stage_buf_);                \
+            MPIDIG_GPU_REQUEST(req, stage_buf) = NULL;                      \
+        }                                                                   \
+    } while (0)
+#else
+#define MPIDIG_GPU_REQUEST_RELEASE(req)
+#endif
+
+MPL_STATIC_INLINE_PREFIX void *MPIDIG_gpu_get_buf(size_t size)
+{
+    void *buf = NULL;
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_GPU_GET_BUF);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_GPU_GET_BUF);
+#ifdef MPL_HAVE_GPU
+    MPIDIU_buf_pool_t **pools = MPIDIG_gpu_global.buf_pool;
+    for (int i = 0; i < MPIDIG_GPU_BUF_POOL_NUM; i++) {
+        if (size > pools[i]->size)
+            continue;
+        buf = MPIDIU_get_buf(pools[i]);
+        goto fn_exit;
+    }
+    /* TODO: for now we simply allocate another buffer and return it
+     * after registering the memory. Ideally, when the buffer is
+     * released it should be added to a huge pool so that it can be
+     * reused later on */
+    MPIDIU_buf_t *tmp_buf;
+    MPL_gpu_malloc_host(&tmp_buf, size + sizeof(MPIDIU_buf_t));
+    tmp_buf->next = NULL;
+    tmp_buf->pool = NULL;
+    buf = tmp_buf->data;
+  fn_exit:
+#endif
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_GPU_GET_BUF);
+    return buf;
+}
+
+MPL_STATIC_INLINE_PREFIX void MPIDIG_gpu_release_buf(void *buf)
+{
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_GPU_RELEASE_BUF);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_GPU_RELEASE_BUF);
+#ifdef MPL_HAVE_GPU
+    MPIDIU_buf_t *cont = MPL_container_of(buf, MPIDIU_buf_t, data);
+    if (cont->pool == NULL) {
+        /* Free huge buffer right away */
+        MPL_gpu_free_host(cont);
+        goto fn_exit;
+    }
+    MPIDIU_release_buf(buf);
+  fn_exit:
+#endif
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_GPU_RELEASE_BUF);
+}
+
+typedef struct MPIDIG_gpu_stage_ctx_t {
+    void *src_buf;
+    size_t size;
+    MPI_Aint true_lb;
+} MPIDIG_gpu_stage_ctx_t;
+
+/* MPIDIG_gpu_stage_send_buf - allocate stage buffer, copy source data into it, and return pointer */
+MPL_STATIC_INLINE_PREFIX void *MPIDIG_gpu_stage_send_buf(const void *buf, MPI_Aint count,
+                                                         MPI_Datatype datatype)
+{
+    char *tmp_buf = (char *) buf;
+#ifdef MPL_HAVE_GPU
+    int dt_contig;
+    size_t size;
+    MPIR_Datatype *dt_ptr;
+    MPI_Aint true_lb;
+
+    MPIDI_Datatype_get_info(count, datatype, dt_contig, size, dt_ptr, true_lb);
+    if (!dt_contig)
+        size = (dt_ptr->extent * count) + (dt_ptr->true_ub - true_lb) - dt_ptr->extent;
```

Comment:
I'm not very clear about datatype, here why we need take out one `extent` size?

Reply:
Also, why can I not use the size returned from `MPIDI_Datatype_get_info`? Try to understand how buf sz is calculated here.

Reply:
The datatype extent might have been resized (i.e. the extent has not be modified using `MPI_Type_create_resized`), therefore we have to add the `true_extent` to and remove the extent to make the size right. If the datatype is not resized then `true_lb = (dt_ptr->true_ub - true_lb)` is equal to `dt_ptr->extent` and you are left with `size = dt_ptr->extent * count`.

Reply:
Sorry I missed the other comment, `MPIDI_Datatype_get_info` returns the number of useful bytes not the extent of the datatype.

Reply:
Okay, it makes sense.

Reply:
BTW I will split this PR into two. Leave this as it is and then create one for GPU utility functions. Eventually this PR will be rebased on top of that.